packages/flutter_tools/gradle/flutter.gradle

```diff
@@ -678,6 +678,26 @@ class FlutterPlugin implements Plugin<Project> {
         project.dependencies.add(configuration, dependency, config)
     }
 
+    // Add a task that can be called on flutter projects that prints the available build variants
+    // in gradle.
+    //
+    // This task prints variants in this format:
+    //
+    // BuildVariant: debug
+    // BuildVariant: release
+    // BuildVariant: profile
+    //
+    // Format of the output of this task is used by `AndroidProject.getBuildVariants`.
+    private static void addTaskForPrintBuildVariants(Project project) {
+        // Warning: The name of this task is used by `AndroidProject.getBuildVariants`.
+        project.tasks.register("printBuildVariants") {
+            description "Prints out all build variants for this Android project"
+            project.android.applicationVariants.all { variant ->
+                println "BuildVariant: ${variant.name}";
+            }
```

Comment:
I believe this will print the output when the task is "configured", which may not mean it's actually going to get run. (See [Gradle docs on the distinction](https://docs.gradle.org/current/userguide/task_configuration_avoidance.html).)

This might be related to why you were seeing the prints happen on `./gradlew tasks --all`: https://github.com/flutter/flutter/pull/123034#discussion_r1151228762
though now that you're using `tasks.register`, I would not expect that to happen. To reproduce the issue, though, it should suffice to [cause this task to be configured](https://docs.gradle.org/current/userguide/task_configuration_avoidance.html#sec:task_configuration_avoidance_pitfalls) — for example, by adding
```groovy
printBuildVariants {
  // This block adds additional configuration to the "printBuildVariants" task.
  // (In this case, it's empty, so it adds no configuration.)
  // But it does so using the eager Gradle API, so as a side effect
  // it causes that task to be configured immediately.
}
```

To fix the issue, wrap the work of the task inside `doLast`, like in [this hello-world example](https://docs.gradle.org/current/userguide/more_about_tasks.html#sec:defining_tasks) from the Gradle docs. That'd look like:
```suggestion
        project.tasks.register("printBuildVariants") {
            description "Prints out all build variants for this Android project"
            doLast {
                project.android.applicationVariants.all { variant ->
                    println "BuildVariant: ${variant.name}";
                }
            }
```

Reply:
Thanks, this indeed fixed the issue.